src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs

```diff
@@ -153,34 +183,28 @@ private IEnumerable<ITagSpan<TTag>> GetTagsForLargeNumberOfSpans(NormalizedSnaps
                     if (currentRequestSpan.Start > currentTagSpan.End)
                     {
                         if (!enumerator.MoveNext())
-                        {
                             break;
-                        }
                     }
                     else if (currentTagSpan.Start > currentRequestSpan.End)
                     {
                         requestIndex++;
 
                         if (requestIndex >= requestedSpans.Count)
-                        {
                             break;
-                        }
                     }
                     else
                     {
-                        if (currentTagSpan.Length > 0)
+                        // Only if this is the first time we are seeing this tag do we add it to the result.
+                        if (currentTagSpan.Length > 0 &&
+                            hashSet.Add(currentTag))
                         {
-                            hashSet.Add(currentTag);
+                            tags.Add(currentTag);
```

Comment:
at the end of the day, teh result of this function was hte hashset itself.  so now, what we do is that if we successfully add to the hash set, that we then place that in the final resultant tags collection.